### PR TITLE
Fix GH Issue 16892 - GC hole due to GT_INDEX_ADDR

### LINKED ISSUE #16892: LinkedList_Generic_Tests_int.RemoveLast_Tests failed on Windows in CI
_From @stephentoub on March 7, 2018 14:54_

https://mc.dot.net/#/user/stephentoub/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/35efdbdc1e25f83f216064482479e555425ce3a0/workItem/System.Collections.Tests/analysis/xunit/System.Collections.Tests.LinkedList_Generic_Tests_int~2FRemoveLast_Tests
```
Windows.81.Amd64.Open-Debug-x64
Unhandled Exception of Type Xunit.Sdk.EqualException
Message :
Assert.Equal() Failure
Expected: LinkedListNode`1 { List = [718949278, 815265269, 911581260, 1007897251, 1104213242, ...], Next = LinkedListNode`1 { List = [...], Next = null, Previous = LinkedListNode`1 { ... }, Value = 1296845224 }, Previous = LinkedListNode`1 { List = [...], Next = LinkedListNode`1 { ... }, Previous = LinkedListNode`1 { ... }, Value = 1104213242 }, Value = 1200529233 }
Actual:   LinkedListNode`1 { List = [718949278, 815265269, 911581260, 1007897251, 1104213242, ...], Next = LinkedListNode`1 { List = [...], Next = null, Previous = LinkedListNode`1 { ... }, Value = 1296845224 }, Previous = LinkedListNode`1 { List = [...], Next = LinkedListNode`1 { ... }, Previous = LinkedListNode`1 { ... }, Value = 1104213242 }, Value = 1200529233 }
Stack Trace :
   at System.Collections.Tests.LinkedList_Generic_Tests`1.VerifyLinkedListNode(LinkedListNode`1 node, T expectedValue, LinkedList`1 expectedList, LinkedListNode`1 expectedPrevious, LinkedListNode`1 expectedNext) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Collections\tests\Generic\LinkedList\LinkedList.Generic.Tests.cs:line 141
   at System.Collections.Tests.LinkedList_Generic_Tests`1.VerifyState(LinkedList`1 linkedList, T[] expectedItems) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Collections\tests\Generic\LinkedList\LinkedList.Generic.Tests.cs:line 83
   at System.Collections.Tests.LinkedList_Generic_Tests`1.InitialItems_Tests(LinkedList`1 collection, T[] expectedItems) in D:\j\workspace\windows-TGrou---74aa877a\src\System.Collections\tests\Generic\LinkedList\LinkedList.Generic.Tests.cs:line 38
   at System.Collections.Tests.LinkedList_Generic_Tests`1.RemoveLast_Tests() in D:\j\workspace\windows-TGrou---74aa877a\src\System.Collections\tests\Generic\LinkedList\LinkedList.Generic.Tests.RemoveLast.cs:line 89
```

_Copied from original issue: dotnet/corefx#27817_

### PR DESCRIPTION
**Pull request by briansull**
Fixes #16892

### CODE REVIEW

**Comment by briansull** [comment]
@AndyAyersMS  PTAL

Also review #13245 for more info

**Comment by briansull** [comment]
I will try to obtain a test case:
The method must be compiled for Debug, 
Have GC args in the outgoing stack area, (i.e as the 5 or 6 argument)
The first four args should be side-effect free (of calls and assignments) and one of the first four args  should  use an array element (with a bounds check operation).

**Comment by briansull** [comment]
@dotnet-bot test Windows_NT x64 Checked gcstress0xc

**Comment by AndyAyersMS** [comment]
Kind of scary that this has been in the code so long without causing trouble.

Do we commonly run GC stress on debug codegen? If not, perhaps we should.

Change LGTM but let's see what Carol / Bruce have to say.

**Comment by danmosemsft** [comment]
I have asked vendors to try running CoreFX tests with GCStress=3 on chk builds as well as the usual ret (at least chk of CoreFX, we don't test chk of CoreCLR over in CoreFX)

Do you guys run GCStress using CoreFX tests at all? Eg in CI?g

**Comment by briansull** [comment]
Yes we can/do run gc stress.  I actually ran it on this PR with the text: (see above)
<AtSign>test Windows_NT x64 Checked gcstress0xc

Note that you need to run gcstress 4 or C to get coverange on every instruction in the JIT code.
GCStress 1 or 3 isn't really going to catch this kind of a GC issue.

Note that using gcstress 4 or C apparently also requires access to a Disassembler library for x64 instructions which wasn't available automatically for me when I tried running it by hand.

**Comment by jkotas** [comment]
> I have asked vendors to try running CoreFX tests with GCStress=3 on chk builds as well as the usual ret  (at least chk of CoreFX, we don't test chk of CoreCLR over in CoreFX)

@danmosemsft I think it is a lot more valuable to run GCStress on CoreCLR checked builds. Checked CoreCLR builds have a lot of important asserts that catch problems sooner, and even in situations that do not lead to crashes.

I have traced down this GC hole using CoreCLR checked build. It resulted into assert on checked CoreCLR build that was hit pretty close to the place where the problem happened. If I was trying to trace it down using CoreCLR release build, I would be probably still scratching my head where it might be coming from.

**Comment by danmosemsft** [comment]
@jkotas is that just better for diagnosing or also for finding? If former then I would rather test on retail so it doesn't take a life age to complete then repeat the run using Chk on whatever set of tests failed. Right now is really  slow on ret and found nothing so far.

**Comment by danmosemsft** [comment]
OK. Also will talk to @valenis. Probably a few random vendor VMs is not going to cut it.

**Comment by jkotas** [comment]
> Right now is really slow on ret and found nothing so far.

It is slow because of the default xunit driver is not suitable for runtime testing. The default xunit driver is using a ton of reflection, and so you are basically GC stress testing the heck out of the one reflection path again and again, and a little bit of the actual tests.

Runtime testing like this needs a different xunit driver that does the reflection at "build time" and transforms the test into a bunch of static calls without reflection. (ProjectN tests that live in TFS use driver like that.)

**Comment by AndyAyersMS** [comment]
I get that you can run gc stress in CI like you did above. Just wondering how much gc stress testing we get for the case where the jit is generating debug code.

I know some JIT tests have `_d` variants so we get a some coverage that way, but the core library and framework assemblies we use in CI are probably not requesting debug codegen.

Looks like we can sort of simulate this via `COMPlus_JitDebuggable` (like we do for minopts) and then running stress on a CHK build. It won't quite have the same IL though.

**Comment by briansull** [comment]
Shall I check this in now?
@AndyAyersMS @CarolEidt  @BruceForstall

**Comment by BruceForstall** [comment]
FWIW, CoreCLR runs lots of CoreFX test jobs, and runs lots of GCStress test jobs, but has no CoreFX + GCStress jobs (e.g., see https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/ for all "stress" jobs). Partially this is because CoreFX jobs are not clean, and GCStress jobs are not clean, and the combination is probably even worse (less likely to run clean).

**Comment by BruceForstall** [comment]
@AndyAyersMS Looks like we currently set debuggable via:
```
    opts.compDbgCode = jitFlags->IsSet(JitFlags::JIT_FLAG_DEBUG_CODE);
```

Maybe we need to add a debuggable stress mode, e.g.

```
STRESS_MODE(DEBUGGABLE)
```

and then use:
```
    opts.compDbgCode = jitFlags->IsSet(JitFlags::JIT_FLAG_DEBUG_CODE) || compStressCompile(STRESS_DEBUGGABLE, 10));
```

**Comment by briansull** [comment]
I added a test case that fails when run under GCStress=4

**Comment by briansull** [comment]
@dotnet-bot test Windows_NT x64 Checked gcstress0xc

File: tests/src/JIT/Regression/JitBlue/GitHub_16892/GitHub_16892.csproj [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+  </PropertyGroup>
+  <ItemGroup>
+    <CodeAnalysisDependentAssemblyPaths Condition=" '$(VS100COMNTOOLS)' != '' " Include="$(VS100COMNTOOLS)..\IDE\PrivateAssemblies">
+      <Visible>False</Visible>
+    </CodeAnalysisDependentAssemblyPaths>
+  </ItemGroup>
+  <PropertyGroup>
+    <DebugType>Full</DebugType>
```
Comment (briansull) [inline tests/src/JIT/Regression/JitBlue/GitHub_16892/GitHub_16892.csproj]: Build this test with Debug==Full

**Comment by briansull** [comment]
@dotnet-bot test Windows_NT x64 Checked gcstress0xc

**Pull request merged**